### PR TITLE
Psycopg2

### PR DESCRIPTION
**Pull request by ocefpaf**
Supersedes #101

### CODE REVIEW

**Comment by msarahan** [comment]
The errors on win are something to do with a weird version of msbuild.  Let's talk about this tomorrow.

**Comment by patricksnape** [comment]
Tried debugging this but for some reason perl on my VM refuses to see `nmake` so  I can't even get to the errors...

**Comment by ocefpaf** [comment]
Thanks for looking into it @patricksnape. I am lost here...

**Comment by msarahan** [comment]
It's definitely something weird going wrong with Perl, not msvc.  Sorry, I'm swamped fixing Unicode issues with our installer, but I hope to get back to fixing recipes soon.

**Comment by ocefpaf** [comment]
See #367

**Comment by jakirkham** [comment]
Restarting CI just to see where we are here relative to the current state of the stack.

**Comment by jseabold** [comment]
Can someone with credentials ssh in to the build box and paste that log file that mentioned in the failure? 

Forgive me as I'm sure I've missed much discussion on this, but why are you building postgres? Could you just work of an existing public postgres container in circle, so we're not failing on the build postgres step? We don't also need to ship postgres with this do we?

**Comment by Korijn** [comment]
@jseabold I know that postgres is at least required for the build of psycopg2, but I also have my doubts about runtime.

@ocefpaf @jakirkham I'm hoping to get this recipe sorted. We currently have an alternate route that works but it's not exactly pretty: https://github.com/ClinicalGraphics/conda-recipes/blob/master/psycopg2/bld.bat perhaps it can be of use to compare our results?

**Comment by Korijn** [comment]
That approach doesn't require full postgres as a dependency at runtime, by the way.

**Comment by ocefpaf** [comment]
@Korijn both @msarahan and I tried here but failed (me lack of knowledge him lack of time). If you want to take this over please summit your recipe and I will gladly close this PR. (We can try `postgres` later.)

**Comment by jseabold** [comment]
AFAICT, cirlce builds are failing on the postgres step. The logs from the circle build would help suggest a fix, so someone needs to SSH in after a failed build and post them. Otherwise, you can install postgres from a repo to build. I don't know about how this would affect compatibility across postgres version.

I haven't looked at the windows builds and don't know how you would do this for appveyor.

**Comment by ocefpaf** [comment]
@msarahan can you review and merge if you are OK with this?

File: recipes/psycopg2/bld.bat [hunk 1]
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install
+if errorlevel 1 exit 1
```
Comment (ocefpaf) [inline recipes/psycopg2/bld.bat]: Leaving this here as I will try to tackle this as soon as the feedstock is created.

**Comment by msarahan** [comment]
I don't know about the runtime postgres  requirement. Can you try removing it?

**Comment by ocefpaf** [comment]
> I don't know about the runtime postgres requirement. Can you try removing it?

Done.

**Comment by msarahan** [comment]
So we need it for libpq.  If @jseabold or @Korijn know a different way to provide it, I'm fine with it. We don't need to provide a full postgres server for psycopg2, I think.

**Comment by ocefpaf** [comment]
> So we need it for libpq. If @jseabold or @Korijn know a different way to provide it, I'm fine with it. We don't need to provide a full postgres server for psycopg2, I think.

Waiting on @jseabold and @Korijn's input before re-adding postgresql to run time requirements.

For google's sake this is the error we got when we removed it:

```shell
===== testing package: psycopg2-2.6.1-py27_0 =====
import: 'psycopg2'
Traceback (most recent call last):
  File "/opt/conda/conda-bld/test-tmp_dir/run_test.py", line 26, in <module>
    import psycopg2
  File "/opt/conda/envs/_test/lib/python2.7/site-packages/psycopg2/__init__.py", line 50, in <module>
    from psycopg2._psycopg import BINARY, NUMBER, STRING, DATETIME, ROWID
ImportError: libpq.so.5: cannot open shared object file: No such file or directory
TESTS FAILED: psycopg2-2.6.1-py27_0
```

**Comment by ocefpaf** [comment]
> Can someone with credentials ssh in to the build box and paste that log file mentioned in the failure?

@jseabold we have the docker image used for the `CircleCI` on dockerhub BTW. You can get it and use to play locally with:

`docker pull condaforge/linux-anvil`

**Comment by Korijn** [comment]
We repacked this windows distribution that doesn't require a full postgresql instance as a runtime dependency: http://www.stickpeople.com/projects/python/win-psycopg/

I recently emailed the author to ask if he was willing to share his build scripts and thankfully, he sent me a rough run through his process. Here it is:

```
Here's a rough rundown on building the dependencies:
ZLib - build all
nmake -f win32/Makefile.msc clean
nmake -f win32/Makefile.msc LOC="-DASMV -DASMINF" OBJA="inffas32.obj match686.obj" zlib.lib

OpenSSL - build all (with command line switches pointing to the zlib include/lib directory.  Note you could probably speed up the process by building only the libraries)
perl Configure zlib --with-zlib-lib=C:/builds/zlib --with-zlib-include=C:/builds/zlib VC-WIN32
CALL ms\do_nasm
nmake -f ms\ntdll.mak vclean
nmake -f ms\nt.mak vclean
nmake -f ms\nt.mak all

PostgreSQL - build libpgcommon libpq (with command line switches pointing to the openssl include/lib directory)
REM PG 9.3+ we need to link against libpgcommon, so build it
cd \builds\postgresql\src\tools\msvc
copy \builds\config.pl .
call clean.bat
perl build.pl libpgport

REM build libpq as a library
cd \builds\postgresql\src\interfaces\libpq
copy \builds\win32.mak .
SET LINK=crypt32.lib

rem check CPU=i386, AMD64
nmake -f win32.mak USE_SSL=1 ENABLE_THREAD_SAFETY=1 SSL_INC=\builds\openssl\inc32  SSL_LIB_PATH=\builds\openssl\out32 clean
nmake -f win32.mak USE_SSL=1 ENABLE_THREAD_SAFETY=1 SSL_INC=\builds\openssl\inc32  SSL_LIB_PATH=\builds\openssl\out32 config .\Release\libpq.lib

Then for psycopg2, I do:
python setup.py clean --all
python setup.py build_ext --have-ssl --pg-config=C:\bin\bin\pg_config -l zlib -L %OSSLLIB%;%LIBPQLIB%;%ZLIBDIR% -I %PGDIR%\src\include;%LIBPQINC% bdist_wininst

Note that C:\bin\bin\pg_config is a binary I grabbed from an install of the PostgreSQL version I'm compiling against.  setup.py uses this program to determine what version of PostgreSQL we are building against and therefor what options to enable.

The variables on the python line are:
OSSLLIB = openssl lib dir
LIBPQLIB = libpq lib directory
ZLIBDIR = zlib lib directory

PGDIR = Postgresql Source dir
LIBPQINC = Postgresql libpq include dir (\src\include\interfaces\libpq)
```

I don't have time to expand my input on this topic further at this point, but I would really prefer not to have a full postgresql instance as a runtime dependency on this package... it's totally impractical for deployment where you're setting up e.g. a django app against a remote postgresql instance. Why would you need a local postgresql instance to come with the package? Or am I misunderstanding here?

**Comment by msarahan** [comment]
Looks great! Thanks! I think we need a different postgres build in light of this,  or at least different packaging. We need the library and the pg_config binary, plus headers I guess.

**Comment by Korijn** [comment]
The `pg_config` binary can be supplied by a full postgres instance. But that's only a build-time requirement.

**Comment by msarahan** [comment]
This is on my plate to fix - I'm planning on creating a "libpq" package that has only the libpq shared library and anything else required at runtime.  I have several other bugs to fix in the meantime, so if anyone wants to take that on, I'm happy to offer tips, guidance, whatever.

**Comment by wlattner** [comment]
That's right, on linux at least, you only need `libpq-dev` which has both `libpq` and `pg_config` rather than building *all* of postgres.

**Pull request closed without merging**